pkg/reconciler/vertex/scaling/scaling.go

```diff
@@ -498,3 +501,44 @@ func (s *Scaler) patchVertexReplicas(ctx context.Context, vertex *dfv1.Vertex, d
 func KeyOfVertex(vertex dfv1.Vertex) string {
 	return fmt.Sprintf("%s/%s", vertex.Namespace, vertex.Name)
 }
+
+func (s *Scaler) getBufferInfos(
+	ctx context.Context,
+	key string,
+	d *daemonclient.DaemonClient,
+	pl *dfv1.Pipeline,
+	vertex *dfv1.Vertex,
+) (
+	partitionBufferLengths []int64,
+	partitionAvailableBufferLengths[]int64,
+	totalBufferLength int64,
+	totalCurrentPending int64,
+	err error,
+) {
+	partitionBufferLengths = make([]int64, 0)
+	partitionAvailableBufferLengths = make([]int64, 0)
+	totalBufferLength = int64(0)
+	totalCurrentPending = int64(0)
+	if !vertex.IsASource() { // Only non-source vertex has buffer to read
+		for _, bufferName := range vertex.OwnedBuffers() {
+			if bInfo, err := d.GetPipelineBuffer(ctx, pl.Name, bufferName); err != nil {
+				err = fmt.Errorf("failed to get the buffer information of vertex %q, %w", vertex.Name, err)
+				return partitionBufferLengths, partitionAvailableBufferLengths, totalBufferLength, totalCurrentPending, err
+			} else {
+				if bInfo.BufferLength == nil || bInfo.BufferUsageLimit == nil || bInfo.PendingCount == nil {
+					err = fmt.Errorf("invalid read buffer information of vertex %q, length, pendingCount or usage limit is missing", vertex.Name)
+					return partitionBufferLengths, partitionAvailableBufferLengths, totalBufferLength, totalCurrentPending, err
+				}
+
+				partitionBufferLengths = append(partitionBufferLengths, int64(float64(bInfo.GetBufferLength())*bInfo.GetBufferUsageLimit()))
+				partitionAvailableBufferLengths = append(partitionAvailableBufferLengths, int64(float64(bInfo.GetBufferLength())*float64(vertex.Spec.Scale.GetTargetBufferAvailability())/100))
+				totalBufferLength += int64(float64(*bInfo.BufferLength) * *bInfo.BufferUsageLimit)
+				totalCurrentPending += *bInfo.PendingCount
+			}
+		}
+
+		_ = s.vertexMetricsCache.Add(key+"/length", totalBufferLength)
```

Comment:
Return the `totalBufferLength`, and then add it to the cache in `scaleOneVertex()`, make it closer to adding `key+"/pending"` to the cache, so that it would be easy to read the code. In that case, change this to a standalone function.